### PR TITLE
[BEAM-7746] Minor typing updates / fixes

### PR DESCRIPTION
**Pull request by chadrik**
This PR corrects a few problems that have occurred over the past few weeks:

-  bad annotations added in PRs
- mangling of type ignore comments by yapf  (fixed up the issues and added an option to the config to prevent it in the future)

Plus it adds some more type annotations.  Nothing too radical or dangerous here. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by chadrik** [comment]
R: @RobbeSneyders 
R: @udim

**Comment by chadrik** [comment]
R: @robertwb

File: sdks/python/apache_beam/io/iobase.py [hunk 1]
```diff
@@ -1289,9 +1289,9 @@ class RestrictionProgress(object):
   """
   def __init__(self, **kwargs):
     # Only accept keyword arguments.
-    self._fraction = kwargs.pop('fraction', None)
-    self._completed = kwargs.pop('completed', None)
-    self._remaining = kwargs.pop('remaining', None)
+    self._fraction = kwargs.pop('fraction', None)  # type: Optional[float]
+    self._completed = kwargs.pop('completed', None)  # type: Optional[int]
```
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: I don't know that we want to restrict this to integral values.

File: sdks/python/apache_beam/runners/common.py [hunk 2]
```diff
@@ -375,7 +379,7 @@ class DoFnInvoker(object):
   represented by a given DoFnSignature."""
 
   def __init__(self,
-               output_processor,  # type: OutputProcessor
+               output_processor,  # type: _OutputProcessor
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Why this change? (And below.)

File: sdks/python/apache_beam/runners/common.py [hunk 3]
```diff
@@ -529,7 +530,7 @@ class SimpleInvoker(DoFnInvoker):
   """An invoker that processes elements ignoring windowing information."""
 
   def __init__(self,
-               output_processor,  # type: OutputProcessor
+               output_processor,  # type: _OutputProcessor
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/common.py]: Same.

File: sdks/python/apache_beam/transforms/external_java.py [hunk 4]
```diff
@@ -37,18 +39,19 @@
 
 # Protect against environments where apitools library is not available.
 # pylint: disable=wrong-import-order, wrong-import-position
+apiclient = None  # type: Optional[types.ModuleType]
```
Comment (robertwb) [inline sdks/python/apache_beam/transforms/external_java.py]: I prefer the previous format, where there was single assignment rather than re-assignment. Is there any benefit to typing this?

File: sdks/python/apache_beam/io/iobase.py (hunk 1 above)
Comment (chadrik) [reply sdks/python/apache_beam/io/iobase.py]: I was a pretty confused by this class, and I think there may actually be some bugs in its handling of `None`-values, so I tried to do the bare minimum of what I thought was right.  No worries. I will revert this and leave this as a final typing task to someone who understands it.

File: sdks/python/apache_beam/transforms/external_java.py (hunk 4 above)
Comment (chadrik) [reply sdks/python/apache_beam/transforms/external_java.py]: I completely agree.  Unfortunately, this is a syntax error in mypy:

```python
try:
  from apache_beam.runners.dataflow.internal import apiclient  # type: Optional[types.ModuleType]
except ImportError:
  apiclient = None
```

As is this:
```python
try:
  import apache_beam.runners.dataflow.internal.apiclient as apiclient  # type: Optional[types.ModuleType]
except ImportError:
  apiclient = None
```

The reason is that type comments are (by design) not capable of doing anything that the new PEP 526 variable annotations are not.  In other words, this is obviously wrong:

```python
try:
  Optional[types.ModuleType]: import apache_beam.runners.dataflow.internal.apiclient as apiclient 
except ImportError:
  apiclient = None
```

In that context, this makes more sense:

```python
apiclient: Optional[types.ModuleType] = None
try:
  from apache_beam.runners.dataflow.internal import apiclient
except ImportError:
  pass
```

File: sdks/python/apache_beam/runners/common.py (hunk 2 above)
Comment (chadrik) [reply sdks/python/apache_beam/runners/common.py]: hmmm... well, at the time that I made this change I _think_ it resolved an error, but either I am mistaken or something changed in the module.   I'm rolling this back.

**Comment by chadrik** [comment]
Down to 79 errors!

Btw, mypy may have revealed some legitimate errors in a recent change:

```
apache_beam/runners/worker/operations.py:837: error: "float" not callable  [operator]
apache_beam/runners/worker/operations.py:838: error: "float" not callable  [operator]
apache_beam/runners/worker/operations.py:839: error: "float" not callable  [operator]
apache_beam/runners/worker/operations.py:841: error: "float" not callable  [operator]
apache_beam/runners/worker/operations.py:842: error: "float" not callable  [operator]
```

File: sdks/python/apache_beam/transforms/external_java.py (hunk 4 above)
Comment (robertwb) [reply sdks/python/apache_beam/transforms/external_java.py]: My question is why we need to type apiclient as types.ModuleType at all.

Comment (chadrik) [reply sdks/python/apache_beam/transforms/external_java.py]: > My question is why we need to type apiclient as types.ModuleType at all.

Fair question.  

If we don't do this, we get the following error:
```
apache_beam/transforms/external_java.py:46: error: Incompatible types in assignment (expression has type "None", variable has type Module)  [assignment]
```

To resolve this, we need to mark `apiclient` as `Optional` 

--- 
why can't we save ourselves some headache and leave out of the `types.ModuleType` part?

```python
apiclient = None  # type: Optional
try:
  from apache_beam.runners.dataflow.internal import apiclient
except ImportError:
  pass
```
 
If we do this, `apiclient` will become `Optional[Any]`

---

Why can't we ignore the error?  

We can, but then mypy will mark the type as non-Optional, and that would remove the added protections that mypy provides against accidentally using the variable when it's `None`.


---

Why can't we just add the type comment on the original `apiclient = None` line? 

```python
try:
  from apache_beam.runners.dataflow.internal import apiclient
except ImportError:
  apiclient = None  # type: Optional[types.ModuleType]
```

With this, we get the following error:

```
apache_beam/transforms/external_java.py:46: error: Name 'apiclient' already defined (by an import)  [no-redef]
```

There is only one opportunity to override/influence the inferred type of a variable:  on the first line where it is defined (think of type variable definitions like C/C++, but with python scoping rules).  However, `apiclient` is defined via an import rather than an assignment, which forces us to preface the import with a variable definition.

**Comment by chadrik** [comment]
Run Python2_PVR_Flink PreCommit

File: sdks/python/apache_beam/transforms/external_java.py (hunk 4 above)
Comment (chadrik) [reply sdks/python/apache_beam/transforms/external_java.py]: I did some more research on this, and I found this mypy issue: https://github.com/python/mypy/issues/1297

It suggests this idiom:

```python
try:
  from apache_beam.runners.dataflow.internal import apiclient as _apiclient
except ImportError:
  apiclient = None
else:
  apiclient = _apiclient
```

The import is a bit longer and uglier, but it has 2 advantages:
- no need to import `Optional` or `ModuleType`
- the idiom I was using was actually making `apiclient` a generic ModuleType, dropping all knowledge of the members of `apache_beam.runners.dataflow.internal`.  That's bad!

The reason this works without explicit `Optional` annotation that mypy will automatically determine optionality in some cases, like this:

```python
if some_conditional():
  x = None
else:
  x = 1
reveal_type(x)  # Revealed type is 'Union[builtins.int, None]'
```

**Comment by chadrik** [comment]
Run Python PreCommit

**Comment by chadrik** [comment]
@robertwb  I don't think the test failures are my fault because they were passing before I rebased onto master...

**Comment by udim** [comment]
LGTM, tests already failing here:
https://builds.apache.org/job/beam_PreCommit_Python_Cron/2443/

**Comment by chadrik** [comment]
Run PythonLint PreCommit

**Comment by chadrik** [comment]
Do we think this is safe to merge?  I've been watching master for something that looks like it could solve the current test problems, and rebasing periodically.

**Comment by chadrik** [comment]
@udim  All tests passing now

**Comment by udim** [comment]
Thanks, Chad! merging

**Comment by angoenka** [comment]
Thanks @chadrik @udim 

Just a quick reminder that we should squash the fixup related commits to keep the history clean based on the committer guide https://beam.apache.org/contribute/committer-guide/#finishing-touches 
Its ok to keep commits separate if they are unrelated changes.

**Comment by udim** [comment]
I want to leave some commits and squash others. There is no way to do that in the GH UI.
